presto-spark-base/src/main/java/com/facebook/presto/spark/planner/optimizers/PickJoinSides.java

```diff
@@ -0,0 +1,117 @@
+import static com.facebook.presto.sql.planner.iterative.rule.DetermineJoinDistributionType.isSmall;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+
+/**
+ * This optimizer chooses the build and probe side of the join based on the size of the
+ * join inputs, putting the smaller table on the right side/ build side of the join.
+ * <p>
+ * for example, consider a join with two children, nodeA and nodeB
+ * if  nodeA is smaller than nodeB, we will convert
+ *         JOIN                       JOIN
+ *        /    \          = >        /    \
+ *    nodeA   nodeB               nodeB   nodeA
+ */
+public class PickJoinSides
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(joinNode ->
+            joinNode.getDistributionType().isPresent()
+                    && joinNode.getDistributionType().get() == PARTITIONED
+                    // Flipping right/left non-equality joins is only supported when
+                    // changing the distribution type too
+                    && !(joinNode.getCriteria().isEmpty() && (joinNode.getType() == LEFT || joinNode.getType() == RIGHT)));
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isAdaptivePickJoinSidesEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        StatsProvider statsProvider = context.getStatsProvider();
+        double leftSize = statsProvider.getStats(joinNode.getLeft()).getOutputSizeInBytes();
+        double rightSize = statsProvider.getStats(joinNode.getRight()).getOutputSizeInBytes();
+
+        if (rightSize > leftSize) {
+            return Result.ofPlanNode(joinNode.flipChildren());
+        }
+
+        // if we don't have exact costs for the join, but based on source tables we think the left side
+        // is very small or much smaller than the right, then flip the join.
+        if (isSizeBasedJoinDistributionTypeEnabled(context.getSession()) && (Double.isNaN(leftSize) || Double.isNaN(rightSize)) && isLeftSideSmall(joinNode, context)) {
+            return Result.ofPlanNode(joinNode.flipChildren());
+        }
+
+        return Result.empty();
+    }
+
+    // This logic is based on DetermineJoinDistributionType.getSizeBasedJoin(),
+    // but does not include switching the join distribution type.
+    private boolean isLeftSideSmall(JoinNode joinNode, Context context)
+    {
+        // the source table of the left side of the join is below the broadcast limit
+        boolean isRightSideSmall = isSmall(joinNode.getRight(), context);
```

Comment:
Why do we have a special case checking for small enough for broadcast limit, when we are doing partitioned joins?

Reply:
it's to match the logic in DetermineJoinDistributionType, where if broadcast join is disabled we still switch sides without broadcasting.

Reply:
also, part of the reasoning for that is because for size based join we're saying "this table is so small that it's probably the one you want to have on the build side, and even if the other table might be even smaller, putting this on the build side would still be fine" so broadcast limit is sort of a proxy for when you consider a table to be very small table.